src/boundary_condition.cpp

```diff
@@ -16,7 +17,18 @@ namespace openmc {
 
 void VacuumBC::handle_particle(Particle& p, const Surface& surf) const
 {
-  p.cross_vacuum_bc(surf);
+  // Random ray and Monte Carlo need different treatments at vacuum BCs
+  if (settings::solver_type == SolverType::RANDOM_RAY) {
+    // Reflect ray off of the surface
+    ReflectiveBC().handle_particle(p, surf);
+
+    // Set ray's angular flux spectrum to vacuum conditions (zero)
+    RandomRay* r = static_cast<RandomRay*>(&p);
```

Comment:
Ideally the `RandomRay` class would inherit from the `GeometryState` rather than the `Particle` class. It would be much nicer if we could break up some of the logic that's currently in the `Particle` and move what we can into the `GeometryState`. I.e., moving through the geometry, crossing surfaces/lattices etc, should more ideally be an abstract functionality that doesn't have anything to do with neutronics. It would be nice if that was in `GeometryState', and then both the `Particle` and `RandomRay` objects could call the base functions and tack on any neutronics tasks they need around them. This may require a bit of refactoring of the MC code, so might be better left to a dedicated PR on that topic? @gridley also mentioned that there were more enhancements coming to the `GeometryState`, so this may already be in the pipeline?

Reply:
Makes sense. One other comment on this particular line: we should probably be using `dynamic_cast` here, no?

Reply:
My thought was that since this is inside the ` if (settings::solver_type == SolverType::RANDOM_RAY)` check, then it is already known for sure that a `RandomRay` object instead of a `Particle` is being used, so a dynamic cast would not be needed. Let me know if a dynamic cast would still be better though.